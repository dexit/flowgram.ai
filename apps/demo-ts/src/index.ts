export * from './types'
export * from './runtime'
export * from './translate'
export * from './validate'
export * from './util'
