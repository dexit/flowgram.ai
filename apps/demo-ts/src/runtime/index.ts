export * from './run'
export * from './RuntimeError'
