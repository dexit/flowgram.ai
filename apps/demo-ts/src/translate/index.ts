export * from './nodes'
export * from './translate'
