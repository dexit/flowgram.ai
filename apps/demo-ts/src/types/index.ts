export * from '@flow-ide-editor/schema'
export * from './nodes'

export class InvalidSchemaError extends Error {}
export class InvalidNodeTypeError extends Error {}
export class InvalidTypeError extends Error {}
export class NotImplementedError extends Error {}

export type FlowArgs = Record<string, unknown>
