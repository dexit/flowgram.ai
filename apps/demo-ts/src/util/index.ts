export * from './log'
export * from './JSBuffer'
