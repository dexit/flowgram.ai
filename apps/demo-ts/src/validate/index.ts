import { FlowNodeSchema } from '@flow-ide-editor/schema'
import { InvalidSchemaError } from '../types'

// TODO use json-schema validation
export function validateOrThrow(node: FlowNodeSchema): void {
  if (!node.definition) throw new InvalidSchemaError()
  if (!node.definition.info) throw new InvalidSchemaError()
  if (!node.definition.info.title) throw new InvalidSchemaError()
}
