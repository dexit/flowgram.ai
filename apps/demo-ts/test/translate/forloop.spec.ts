// import _ from 'lodash'
import _ from 'lodash'
import { FlowNodeSchema, translateNode } from '../../src'

const variable1: FlowNodeSchema = {
  id: 'variable_88d5e7cdeb6',
  type: 'variable',
  definition: {
    info: {
      title: 'assign',
    },
    outputs: { properties: {}, title: 'Output', type: 'object' },
    script: { language: 'python', content: '' },
  },
  data: {
    setVariables: {
      '0': {
        left: { type: 'ref', content: '$.start.outputs.init' },
        right: {
          type: 'expression',
          content: 'start.outputs.init + loop.localVariables.index',
        },
        operator: 'assign',
      },
    },
  },
}

const forloopMin = {
  id: 'forloop_b288d5e7cde',
  type: 'forloop',
  definition: {
    info: {
      title: 'loop',
    },
    outputs: { title: 'Output', type: 'object', flow: { index: 0 } },
    script: { language: 'python', content: '', mode: 'default' },
    localVariables: {
      type: 'object',
      properties: { index: { type: 'integer', flow: { index: 0 } } },
    },
  },
  data: {
    initializer: { index: { type: 'expression', content: '1' } },
    condition: { type: 'expression', content: 'loop.localVariables.index <= 3' },
    step: { type: 'expression', content: 'loop.localVariables.index += 1' },
  },
  blocks: [],
}

const forloop2 = {
  id: 'forloop_b288d5e7cde',
  type: 'forloop',
  definition: {
    info: {
      title: 'loop',
    },
    outputs: { title: 'Output', type: 'object', flow: { index: 0 } },
    script: { language: 'python', content: '', mode: 'default' },
    localVariables: {
      type: 'object',
      properties: {
        left: { type: 'integer', flow: { index: 0 } },
        right: { type: 'integer', flow: { index: 0 } },
      },
    },
  },
  data: {
    initializer: {
      left: { type: 'expression', content: '1' },
      right: { type: 'expression', content: '10' },
    },
    condition: {
      type: 'expression',
      content: 'loop.localVariables.left <= loop.localVariables.right',
    },
    step: {
      type: 'expression',
      content: 'loop.localVariables.left += 1, loop.localVariables.right -= 1',
    },
  },
  blocks: [],
}

const break0 = {
  id: 'break_4c661c24008',
  type: 'break',
  definition: {
    info: {
      title: 'Break',
    },
    script: { language: 'python', content: '', mode: 'default' },
  },
}

describe('translate', () => {
  it('Condition - min', () => {
    const node = _.cloneDeep(forloopMin) as FlowNodeSchema

    expect(translateNode(node)).toMatchSnapshot()
  })

  it('Forloop - basic', () => {
    const node = _.cloneDeep(forloopMin) as FlowNodeSchema
    node.blocks?.push(_.cloneDeep(variable1))

    expect(translateNode(node)).toMatchSnapshot()
  })

  it('Forloop - multi initilizers', () => {
    const node = _.cloneDeep(forloop2) as FlowNodeSchema

    expect(translateNode(node)).toMatchSnapshot()
  })

  it('Forloop - break', () => {
    const node = _.cloneDeep(forloopMin) as FlowNodeSchema
    node.blocks?.push(_.cloneDeep(break0))

    expect(translateNode(node)).toMatchSnapshot()
  })

  it('Forloop - nest', () => {
    const node = _.cloneDeep(forloopMin) as FlowNodeSchema
    const node1 = _.cloneDeep(forloopMin) as FlowNodeSchema

    node1.blocks?.push(_.cloneDeep(forloopMin) as FlowNodeSchema)
    node.blocks?.push(node1)

    expect(translateNode(node)).toMatchSnapshot()
  })
})
