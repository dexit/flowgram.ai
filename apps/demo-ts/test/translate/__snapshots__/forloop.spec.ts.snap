// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translate Condition - min 1`] = `
"  //=== loop@forloop | {level:1,blockIndex:0}
  const loop = {}
  loop.outputs = {}
  loop.localVariables = {
    index: 0,
  }
  for (loop.localVariables.index = 1; loop.localVariables.index <= 3; loop.localVariables.index += 1) {
  } // loop@forloop
  "
`;

exports[`translate Forloop - basic 1`] = `
"  //=== loop@forloop | {level:1,blockIndex:0}
  const loop = {}
  loop.outputs = {}
  loop.localVariables = {
    index: 0,
  }
  for (loop.localVariables.index = 1; loop.localVariables.index <= 3; loop.localVariables.index += 1) {
    //=== assign@variable | {level:2,blockIndex:0,parent:'loop'}
    const assign = {}
    assign.outputs = {}
    start.outputs.init = start.outputs.init + loop.localVariables.index
    } // loop@forloop
  "
`;

exports[`translate Forloop - break 1`] = `
"  //=== loop@forloop | {level:1,blockIndex:0}
  const loop = {}
  loop.outputs = {}
  loop.localVariables = {
    index: 0,
  }
  for (loop.localVariables.index = 1; loop.localVariables.index <= 3; loop.localVariables.index += 1) {
    //=== Break@break | {level:2,blockIndex:0,parent:'loop'}
    break
    } // loop@forloop
  "
`;

exports[`translate Forloop - multi initilizers 1`] = `
"  //=== loop@forloop | {level:1,blockIndex:0}
  const loop = {}
  loop.outputs = {}
  loop.localVariables = {
    left: 0,
    right: 0,
  }
  for (loop.localVariables.left = 1,loop.localVariables.right = 10; loop.localVariables.left <= loop.localVariables.right; loop.localVariables.left += 1, loop.localVariables.right -= 1) {
  } // loop@forloop
  "
`;

exports[`translate Forloop - nest 1`] = `
"  //=== loop@forloop | {level:1,blockIndex:0}
  const loop = {}
  loop.outputs = {}
  loop.localVariables = {
    index: 0,
  }
  for (loop.localVariables.index = 1; loop.localVariables.index <= 3; loop.localVariables.index += 1) {
    //=== loop@forloop | {level:2,blockIndex:0,parent:'loop'}
    const loop = {}
    loop.outputs = {}
    loop.localVariables = {
      index: 0,
    }
    for (loop.localVariables.index = 1; loop.localVariables.index <= 3; loop.localVariables.index += 1) {
      //=== loop@forloop | {level:3,blockIndex:0,parent:'loop'}
      const loop = {}
      loop.outputs = {}
      loop.localVariables = {
        index: 0,
      }
      for (loop.localVariables.index = 1; loop.localVariables.index <= 3; loop.localVariables.index += 1) {
      } // loop@forloop
      } // loop@forloop
    } // loop@forloop
  "
`;
